azext_iot/tests/central/__init__.py

```diff
@@ -33,18 +47,75 @@
 class CentralLiveScenarioTest(CaptureOutputLiveScenarioTest):
     def __init__(self, test_scenario):
         super(CentralLiveScenarioTest, self).__init__(test_scenario)
-
-    def cmd(self, command, api_version=None, checks=None, expect_failure=False):
-        command = self._appendOptionalArgsToCommand(
-            command, token=TOKEN, dns_suffix=DNS_SUFFIX, api_version=api_version
-        )
+        self._create_app()
+        self.storage_container = STORAGE_CONTAINER
+        self.storage_cstring = STORAGE_CSTRING
+        self.storage_account_name = STORAGE_NAME
+        self.token = None
+        self.dns_suffix = None
+
+    def cmd(
+        self,
+        command,
+        api_version=None,
+        checks=None,
+        expect_failure=False,
+        include_opt_args=True
+    ):
+        if include_opt_args:
+            command = self._appendOptionalArgsToCommand(
+                command, api_version=api_version
+            )
         return super().cmd(command, checks=checks, expect_failure=expect_failure)
 
-    def cmd_withoutParams(self, command, checks=None, expect_failure=False):
-        return super().cmd(command, checks=checks, expect_failure=expect_failure)
+    def _create_app(self):
+        """
+        Create an Iot Central Application if a name is not given in the pytest configuration.
+
+        Will populate the following variables based on given pytest configuration:
+          - app_primary_key
+          - token
+          - dns_suffix
+        """
+        self.app_id = settings.env.azext_iot_central_app_id or "test-app-" + generate_generic_id()
+        self.app_rg = APP_RG
+        # If not populated here, will be populated with get_app_scope_id
+        self._scope_id = settings.env.azext_iot_central_scope_id
+
+        # only populate these if given in the test configuration variables
+        self.app_primary_key = settings.env.azext_iot_central_primarykey
+        self.dns_suffix = settings.env.azext_iot_central_dns_suffix
+        self.token = settings.env.azext_iot_central_token
+
+        # Create Central App if it does not exist. Note that app_primary_key will be nullified since
+        # there is no current way to get the app_primary_key and not all tests can be executed.
+        if not settings.env.azext_iot_central_app_id:
+            if not APP_RG:
+                raise Exception("Tests need either app name or resource group.")
+
+            self.cmd(
+                "iot central app create -n {} -g {} -s {} --mi-system-assigned -l {}".format(
+                    self.app_id,
+                    self.app_rg,
+                    self.app_id,
+                    "westus"
+                ),
+                include_opt_args=False
+            )
+            self.app_primary_key = None
```

Comment:
Once app was created, the self.dns_suffix should be the fixed with "azureiotcentral.com", self.token should not be read from env as well.

Reply:
clean self._scope_id as well.

Reply:
fixing to "azureiotcentral.com" prevents internal development against dev clusters. This is useful when working on changes that require backend features not yet available in Central PROD

Reply:
unless we assume internal dev will always have an existing app and can go with variables in pytest

Reply:
not so sure if ```az iot central app create``` can create an PPE app which had "azureiotcentral-ppe.com", in create app flow, I feel this should be picked up from the app, because we had mooncake app as well.

Reply:
dns_suffix defaults to azureiotcentral.com already for commands so there is no need to set it 
![image](https://user-images.githubusercontent.com/73560279/157304013-f92701ae-5a2b-4df2-9a49-a7f4baa20430.png)

Also will fix the help for this param...

How do you currently create a ppe app or mooncake app? Is it possible to map a location/subscription to a ppe environment (ex: creating an app in X location in Y subscription will always create a ppe app)?
If so, is it possible to set default tokens/scope_id/other variables? For example, all apps created in ppe environment share the scope id? I want to make sure that token, dns_suffix, and scope_id can be safely nulled out if a new app is created.

Reply:
Also regarding the test runs, see the screenshots above - they include time differences between generating apps and not generating apps.

Reply:
updated the tests to null out scope_id too so the scope_id could be taken from the newly created app.  WIll not null out dns_suffix or token for now.

Reply:
Sounds good to me.

Reply:
added token tests